File: frontend/src/utils/__tests__/listingPreferences.test.ts
import { describe, expect, it } from "vitest";
import type { ListingPreferences } from "@/types/user";
import {
  buildListingSections,
  defaultListingPreferences,
  matchPrefixRule,
  normalizeListingPreferences,
  paginateListingSections,
  validatePrefix,
} from "../listingPreferences";

const item = (name: string, isDir: boolean) => ({
  name,
  isDir,
  path: `/${name}`,
});

describe("listing preferences", () => {
  it("migrates legacy dotfile visibility and defaults every group to expanded", () => {
    const visible = defaultListingPreferences(false);
    const hidden = defaultListingPreferences(true);

    expect(visible.prefixRules.map((rule) => rule.prefix)).toEqual([
      ".",
      "@",
      "#",
      "~",
      "$",
    ]);
    expect(visible.prefixRules[0].visible).toBe(true);
    expect(hidden.prefixRules[0].visible).toBe(false);
    expect(hidden.prefixRules.every((rule) => rule.expanded)).toBe(true);
  });

  it("uses the longest matching prefix", () => {
    const preferences: ListingPreferences = {
      version: 1,
      prefixRules: [
        { prefix: "@", visible: true, expanded: true, order: 0 },
        { prefix: "@@", visible: false, expanded: false, order: 1 },
      ],
    };
    expect(matchPrefixRule("@@cache", preferences.prefixRules)?.prefix).toBe(
      "@@"
    );
  });

  it("groups before pagination and excludes hidden or collapsed items", () => {
    const preferences = normalizeListingPreferences({
      version: 1,
      prefixRules: [
        { prefix: ".", visible: false, expanded: true, order: 0 },
        { prefix: "@", visible: true, expanded: false, order: 1 },
        { prefix: "#", visible: true, expanded: true, order: 2 },
      ],
    });
    const sections = buildListingSections(
      [item(".cache", true), item("@system", true), item("normal", true)],
      [item("#notes", false), item("readme", false)],
      preferences
    );

    expect(sections.map((section) => section.id)).toEqual([
      "prefix:@",
      "prefix:#",
      "directories",
      "files",
    ]);
    expect(sections[0].items.map((entry) => entry.name)).toEqual(["@system"]);
    expect(sections[0].expanded).toBe(false);

    const paged = paginateListingSections(sections, 2);
    expect(paged.map((section) => [section.id, section.items.length])).toEqual([
      ["prefix:@", 0],
      ["prefix:#", 1],
      ["directories", 1],
    ]);
  });

  it("validates custom prefixes", () => {
    expect(validatePrefix("@@")).toBeNull();
    expect(validatePrefix("a/b")).not.toBeNull();
    expect(validatePrefix(" ")).not.toBeNull();
    expect(validatePrefix("123456789")).not.toBeNull();
  });
});

File: backend/users/listing_preferences_test.go
package users

import "testing"

func TestDefaultListingPreferencesMigratesLegacyDotfiles(t *testing.T) {
	visible := DefaultListingPreferences(false)
	hidden := DefaultListingPreferences(true)

	if visible.Version != ListingPreferencesVersion || hidden.Version != ListingPreferencesVersion {
		t.Fatalf("versions = visible:%d hidden:%d", visible.Version, hidden.Version)
	}
	if len(visible.PrefixRules) != 5 || len(hidden.PrefixRules) != 5 {
		t.Fatalf("rules = visible:%#v hidden:%#v", visible.PrefixRules, hidden.PrefixRules)
	}
	if !visible.PrefixRules[0].Visible || hidden.PrefixRules[0].Visible {
		t.Fatalf("dot visibility = visible:%t hidden:%t", visible.PrefixRules[0].Visible, hidden.PrefixRules[0].Visible)
	}
	for _, preferences := range []ListingPreferences{visible, hidden} {
		for _, rule := range preferences.PrefixRules {
			if !rule.Expanded {
				t.Fatalf("default rule collapsed: %#v", rule)
			}
		}
	}
}

func TestNormalizeListingPreferencesValidatesAndCompletesRules(t *testing.T) {
	preferences, err := NormalizeListingPreferences(ListingPreferences{
		Version: ListingPreferencesVersion,
		PrefixRules: []PrefixRule{
			{Prefix: "@@", Visible: false, Expanded: false, Order: 8},
			{Prefix: "@", Visible: true, Expanded: true, Order: 3},
		},
	}, false)
	if err != nil {
		t.Fatal(err)
	}
	if len(preferences.PrefixRules) != 6 {
		t.Fatalf("rules = %#v", preferences.PrefixRules)
	}
	matched, ok := preferences.MatchPrefixRule("@@cache")
	if !ok || matched.Prefix != "@@" || matched.Visible || matched.Expanded {
		t.Fatalf("longest match = %#v, %t", matched, ok)
	}

	invalid := []ListingPreferences{
		{Version: 2},
		{Version: ListingPreferencesVersion, PrefixRules: []PrefixRule{{Prefix: ""}}},
		{Version: ListingPreferencesVersion, PrefixRules: []PrefixRule{{Prefix: "a/b"}}},
		{Version: ListingPreferencesVersion, PrefixRules: []PrefixRule{{Prefix: " "}}},
		{Version: ListingPreferencesVersion, PrefixRules: []PrefixRule{{Prefix: "@"}, {Prefix: "@"}}},
	}
	for index, value := range invalid {
		if _, err := NormalizeListingPreferences(value, false); err == nil {
			t.Fatalf("invalid case %d accepted: %#v", index, value)
		}
	}
}

func TestNormalizeListingPreferencesLimitsCustomRules(t *testing.T) {
	rules := make([]PrefixRule, 0, MaxCustomPrefixRules+1)
	for index := 0; index <= MaxCustomPrefixRules; index++ {
		rules = append(rules, PrefixRule{Prefix: string(rune(0x4e00 + index)), Visible: true, Expanded: true, Order: index})
	}
	_, err := NormalizeListingPreferences(ListingPreferences{Version: ListingPreferencesVersion, PrefixRules: rules}, false)
	if err == nil {
		t.Fatal("custom prefix limit was not enforced")
	}
}
